hasktorch/src/Torch/NN.hs

```diff
@@ -41,73 +41,62 @@ nextParameter = do
     [] -> error "Not enough parameters supplied to replaceParameters"
     (p : t) -> do put t; return p
 
-class HasForward f a b | f a -> b where
-  forward :: f -> a -> b
+class HasForward f a where
+  type B f a :: Type
+  type B f a = GB (Rep f) (Rep a)
+
+  forward :: f -> a -> B f a
   default forward ::
     ( Generic f,
       Generic a,
-      Generic b,
-      GHasForward (Rep f) (Rep a) (Rep b)
-    ) =>
-    f ->
-    a ->
-    b
-  forward f a = to $ gForward (from f) (from a)
-  forwardStoch :: f -> a -> IO b
-  default forwardStoch ::
-    ( Generic f,
-      Generic a,
-      Generic b,
-      GHasForward (Rep f) (Rep a) (Rep b)
+      GHasForward (Rep f) (Rep a),
+      B f a ~ GB (Rep f) (Rep a)
     ) =>
     f ->
     a ->
-    IO b
-  forwardStoch f a = to <$> gForwardStoch (from f) (from a)
+    B f a
+  forward f a = gForward (from f) (from a)
 
-class GHasForward (f :: Type -> Type) (a :: Type -> Type) (b :: Type -> Type) | f a -> b where
-  gForward :: forall c c' c''. f c -> a c' -> b c''
-  gForwardStoch :: forall c c' c''. f c -> a c' -> IO (b c)
+class GHasForward (f :: Type -> Type) (a :: Type -> Type) where
+  type GB f a :: Type
+  gForward :: forall c c'. f c -> a c' -> GB f a
```

Comment:
Idea: 
```suggestion
  gForward :: forall c c'. f c -> a c' -> Rep (GB f a)
```
Have you tried that?
I think that way you wouldn't need to work with `()`, `(,)`, and `Either` below.

Reply:
Thanks, I will try it! I think this is why I was trying to make `GB f a :: Type -> Type` before; I hadn't thought of this way. Still barely understand Generics, but slowly improving :smiley:

Reply:
I have been thinking about this some more in the last half an hour, and I'm not so sure any more about my suggestion.
The issue seems to be that the output type is only defined modulo isomorphism.
I am actually not sure why the compiler accepted the fundeps in `master`, because the output type doesn't seem unambiguously defined jointly by `f` and `a`. I suspect that it will complains once one wants to define instances.

I thus think your draft solution is fine. You've got to commit to specific output types, and these may as well be `()`, `(,)`, and `Either`.

Reply:
When I do it this way I'm getting confused about how to implement the type family for e.g. the trivial instance:
```haskell
instance GHasForward U1 U1 where
  type GB U1 U1 = _  -- what goes here?
  gForward U1 U1 = U1
```

Reply:
```Haskell
type GB U1 U1 = _  -- what goes here?
```
Yes, exactly, I realized that, too. You could write `()` of course but that wouldn't improve things.

Reply:
@mcwitt It's late, and I need to sleep over this, but I think we should rethink all of these generic instances. It's probably better to start again from what we are trying to accomplish here.
One thing I came up with just now is Generic1, that is, generics of type constructors...
We could then assume that the type constructor for model, input, and output is the same...

Reply:
Ok, I'm more awake now, let's try to go back to what we want to achieve here:

I'm envisioning automatic derivation capabilities for the following situations:

1) sum types:
```haskell
data Model ~ Either ModelA ModelB  -- using '~' here to indicate that this can be any ADT isomorphic to 'Either'
data In ~ Either InA InB
data Out ~ Either OutA OutB

instance
  (HasForward ModelA InA, OutA ~ B ModelA InA, HasForward ModelB InB, OutB ~ B ModelB InB) =>
  HasForward Model In where
  type B Model In = Maybe Out
  forward (Left modelA) (Left inA) = Just . Left $ forward modelA inA
  forward (Right modelB) (Right inB) = Just . Right $ forward modelB inB
  forward _ _ = Nothing
```
To achieve this, I think we should not only demand that `In` and `Out` are isomorphic, but also that they are the same type constructors. That is, if `In a b` is `Either a b`, then `Out a b` must also be `Either a b`. If `In a b` is
```haskell
data In a b = This { unThis :: a } | That { unThat :: b }
```
then
```haskell
type Out a b = In a b
```
Note that `Model` can be any other ADT that is isomorphic with `Either`. It could be `Either ModelA ModelB` or it could be something else like
```
data OneOfTwoModels a b = ThisModel { unThisModel :: a } | ThatModel {unThatModel :: b }
```
but it mustn't be
```
data OneOfThreeModels a b c = ThisModel { unThisModel :: a } | ThatModel {unThatModel :: b } | TheirModel {unTheirModel :: c }
```

2) product types:
```haskell
data Model ~ (ModelA, ModelB)  -- using '~' here to indicate that this can be any ADT isomorphic to '(,)'
data In ~ (InA, InB)
data Out ~ (OutA, OutB)

instance
  (HasForward ModelA InA, OutA ~ B ModelA InA, HasForward ModelB InB, OutB ~ B ModelB InB) =>
  HasForward Model In where
  type B Model In = Out
  forward (modelA, modelB) (inA, inB) = (forward modelA inA, forward modelB inB)
```
Again, to achieve this, I think we should not only demand that `In` and `Out` are isomorphic, but also that they are the same type constructors. That is, if `In a b` is `(a, b)`, then `Out a b` must also be `(a, b)`. If `In a b` is
```haskell
data In a b = Both { this :: a, that :: b }
```
then
```haskell
type Out a b = In a b
```
Similar considerations as above apply to `Model`.

3) composite types (something that has a `Compose f g` instance where `f` is a functor)
We'll see if this is useful, but consider:
```Haskell
data MaybeModel ~ Maybe Model
data MaybeIn ~ Maybe In
data MaybeOut ~ Maybe Out

instance
  (HasForward Model In, Out ~ B Model In) =>
  HasForward MaybeModel MaybeIn where
  type B Model In = MaybeOut
  forward (Just model) (Just in) = Just $ forward model in
  forward _ _ = Nothing
```

4) lesser variations or combinations of the above
Like:
```haskell
data Model ~ Either ModelA ModelB  -- using '~' here to indicate that this can be any ADT isomorphic to 'Either'
data In = ...
data Out = ...

instance
  (HasForward ModelA In, Out ~ B ModelA In, HasForward ModelB In, Out ~ B ModelB In) =>
  HasForward Model In where
  type B Model In = Out
  forward (Left modelA) in = forward modelA in
  forward (Right modelB) in = forward modelB in
```
and so on. There are many such combinations.

I am quite confident that all of this can be achieved with generics, and I think we should look into `Generic1` (https://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-Generics.html#g:16), because `Generic1` makes it possible to reason about type constructors like `Either a b`.

Reply:
About non-determinism/stochasticity in this context:

If we have `StochOut = G -> (Out, G)` with `G` some random state ADT, we may need to wire the instances differently. Note also that `(->) a b` doesn't have a `Generic` instance (AFAIK).

Let's just look at product types for now. We will have:
```haskell
data Model ~ (ModelA, ModelB)  -- using '~' here to indicate that this can be any ADT isomorphic to '(,)'
data In ~ (InA, InB)
data StochOutA ~ G -> (OutA, G)
data StochOutA ~ G -> (OutB, G)
data StochOut ~ G -> ((OutA, OutB), G)

instance
  (HasForward ModelA InA, StochOutA ~ B ModelA InA, HasForward ModelB InB, StochOutB ~ B ModelB InB) =>
  HasForward Model In where
  type B Model In = StochOut
  forward (modelA, modelB) (inA, inB) =
    \g -> case (forward modelA inA) g of (outA, g') -> case (forward modelB inB) g' of (outB, g'') -> ((outA, outB), g'')
```

Reply:
Ok, I think I know now how to get the above dispatching working for type class instances. We can use a closed typed family of the following form to compute an auxiliary type tag:
```Haskell
data ModelRandomness = Deterministic | Stochastic

type family ModelRandomnessR (out :: Type) :: ModelRandomness where
  ModelRandomnessR ((->) g (_, g)) = 'Stochastic  -- not sure yet if this pattern match is syntactically correct, we'll see...
  ModelRandomNessR _ = 'Deterministic
```
Once we have this, we can use the `ModelRandomness` type-level flag in the following way (for example):
```Haskell
class HasForwardProduct (modelARandomness :: ModelRandomness) (modelBRandomness :: ModelRandomness) f1 a1 f2 a2 where
  type BProduct f1 a1 f2 a2 :: Type
  forwardProduct :: f1 -> a1 -> f2 -> a2 -> BProduct f1 a1 f2 a2

instance
  (HasForward ModelA InA, OutA ~ B ModelA InA, HasForward ModelB InB, OutB ~ B ModelB inB) =>
  HasForwardProduct 'Deterministic 'Deterministic ModelA InA ModelB inB where
  type BProduct ModelA InA ModelB inB = (OutA, OutB)
  forwardProduct modelA inA modelB inB = (forward modelA inA, forward modelB inB)

instance
  (HasForward ModelA InA, OutA ~ B ModelA InA, HasForward ModelB InB, StochOutB ~ B ModelB inB) =>
  HasForwardProduct 'Stochastic 'Deterministic ModelA InA ModelB inB where
  type BProduct ModelA InA ModelB inB = G -> ((OutA, OutB), G)
  forwardProduct modelA inA modelB inB =
    \g -> case (forward modelA inA) g of (outA, g') -> ((outA, forward modelB inB), g')

...
```
plus an instance for `HasForward` for the product type that dispatches to `HasForwardProduct` by means of `ModelRandomnessR`. Will write an example tomorrow.

Reply:
And here's how the dispatching would work:
```Haskell
instance
  ( ModelRandomnessR OutA ~ modelARandomness,
    ModelRandomnessR OutB ~ modelBRandomness,
    HasForwardProduct modelARandomness modelBRandomness ModelA InA ModelB InB
  ) =>
  HasForward Model In where
  type B Model In = BProduct (ModelRandomnessR OutA) (ModelRandomnessR OutB) ModelA InA ModelB InB
  forward (modelA, modelB) (inA, inB) = forwardProduct (Proxy :: Proxy modelARandomness) (Proxy :: Proxy modelBRandomness) modelA inA modelB inB
```

Reply:
I think this should work, provided that the pattern match for the `->` type constructor in the `ModelRandomnessR` type family compiles.

@mcwitt can you see if this makes sense?

If so, then what remains to be done is to generalize these concrete implementations with generics. That is, instead of having `Either`, `(,)`, `Compose`, etc., we need to work with `:+:`, `:*:`, `:.:` and so on. I also don't know if we need `Generic1` or not, but I suspect we do.

Reply:
Awesome! I'll have some time this evening to try to work through it, and will let you know if I have any questions.

Reply:
btw, I tried `ModelRandomnessR`, and it actually works:
```Haskell
data ModelRandomness = Deterministic | Stochastic

data G = G

type family ModelRandomnessR (out :: Type) :: ModelRandomness where
  ModelRandomnessR ((->) G (_, G)) = 'Stochastic
  ModelRandomnessR _ = 'Deterministic
```

```Haskell
> :kind! ModelRandomnessR (G -> (Tensor, G))
ModelRandomnessR (G -> (Tensor, G)) :: ModelRandomness
= 'Stochastic
> :kind! ModelRandomnessR Tensor
ModelRandomnessR Tensor :: ModelRandomness
= 'Deterministic
> :kind! ModelRandomnessR (Int -> (Tensor, Int))
ModelRandomnessR (Int -> (Tensor, Int)) :: ModelRandomness
= 'Deterministic
```